torch/testing/_internal/distributed/rpc/rpc_test.py

```diff
@@ -3551,16 +3551,14 @@ def _test_remote_message_delay_timeout(self, func, args, dst=None):
                 # to_here() should raise timeout error, since it does not know about the
                 # status of rpc.remote().
                 slow_rref.to_here(0.001)
-        # Note: UserRRef will try to send out a RRefUserDelete, but will not find it on the owner.
-        # later, the owner will process the RRef creation, and time out
-        # waiting for the delete messages.
-        # Therefore, we wait until we get notification that pending owners have been confirmed
-        # before sending out RRefUserDeletes.
+        # Note: If we proceed with shutdown, UserRRef will send out a RRefUserDelete
+        # but this can be a noop since it may not exist on the owner yet. Later,
+        # the owner can process the RRef creation and wait for the delete message,
+        # thus leading to a timeout.
+        # Therefore, we wait until we get notification that pending owners have
+        # been confirmed before sending out RRefUserDeletes.
         if dst_rank != self.rank:
             wait_until_owners_and_forks_on_rank(2, 2, rank=dst_rank)
-        else:
-            # 1 owner, fork deleted
-            wait_until_owners_and_forks_on_rank(1, 0, rank=dst_rank)
```

Comment:
Question, by "the `finishCreatingOwnerRRef` callback would run before the `request_callback_impl` finishes processing the request", the request here is not the original `rpc.remote()` request, instead it is the request sent by `wait_until_owners_and_forks_on_rank`? Is this correct?

Reply:
It's actually the `rpc.remote()` request. What happens in this test is that rank0 sends `rpc.remote()` to rank0, but we inject failure so the future times out in 1ms. Since the future is marked with an error, callbacks can be run at any time by the thread that set the errors. 

In another thread, the rpc.remote() is still being processed (we marked it with an error but that does not stop the processing on the receiver). There are 2 cases:

1. Callbacks are run before the request is processed. In this case, we call `finishCreatingOwnerRRef` and remove the owner. Then, rpc.remote() sees that the owner does not exist and creates a new `ownerRRef`. Then there will be 1 owner on the node (this is what the test previously assumed would happen)

2. Callbacks are run after the request is processed. In this case, rpc.remote() used the existing ownerRRef, and in the callback we will remove it, so there will be 0 owners (this fix should make this case work).

Reply:
I see. I wonder if we should handle this error more explicitly. For the rpc-to-self case, if the receiving end expects the `OwnerRRef` to be there before processing the request, can we change the following two lines and guard them by checking `rrefId != forkId`? If it is rpc-to-self, and the OwnerRRef does not exist, shall we just let it crash?

https://github.com/pytorch/pytorch/blob/53af9df557aff745edf24193ece784fd008c6f19/torch/csrc/distributed/rpc/request_callback_impl.cpp#L318

https://github.com/pytorch/pytorch/blob/53af9df557aff745edf24193ece784fd008c6f19/torch/csrc/distributed/rpc/request_callback_impl.cpp#L429

Reply:
Added this in the latest version.